edb/server/protocol/auth_ext/http.py

```diff
@@ -818,12 +829,178 @@ async def handle_reset_password(self, request: Any, response: Any):
+            data,
+            {"provider", "email", "challenge", "callback_url"},
+        )
+
+        email = data["email"]
+        challenge = data["challenge"]
+        callback_url = data["callback_url"]
+        if not self._is_url_allowed(callback_url):
+            raise errors.InvalidData(
+                "Callback URL does not match any allowed URLs.",
+            )
+        maybe_redirect_to = data.get("redirect_to")
+        if maybe_redirect_to and not self._is_url_allowed(maybe_redirect_to):
+            raise errors.InvalidData(
+                "Redirect URL does not match any allowed URLs.",
+            )
+
+        magic_link_client = magic_link.Client(
+            db=self.db,
+            issuer=self.base_path,
+            tenant=self.tenant,
+            test_mode=self.test_mode,
+        )
+        try:
+            await magic_link_client.send_magic_link(
+                email=email,
+                callback_url=callback_url,
+                link_url=f"{self.base_path}/magic-link/authenticate",
+                challenge=challenge,
+            )
+
+            return_data = {
+                "email_sent": email,
+            }
+
+            if maybe_redirect_to:
+                response.status = http.HTTPStatus.FOUND
+                response.custom_headers["Location"] = util.join_url_params(
+                    maybe_redirect_to, return_data
+                )
+            else:
+                response.status = http.HTTPStatus.OK
+                response.content_type = b"application/json"
+                response.body = json.dumps(return_data).encode()
+        except Exception as ex:
+            redirect_on_failure = data.get(
+                "redirect_on_failure", maybe_redirect_to
+            )
+            if redirect_on_failure is None:
+                raise ex
+            else:
+                if not self._is_url_allowed(redirect_on_failure):
+                    raise errors.InvalidData(
+                        "Redirect URL does not match any allowed URLs.",
+                    )
+                response.status = http.HTTPStatus.FOUND
+                redirect_params = {
+                    "error": str(ex),
+                    "email": data.get('email', ''),
+                }
+                response.custom_headers["Location"] = util.join_url_params(
+                    redirect_on_failure, redirect_params
+                )
+
+    async def handle_magic_link_authenticate(self, request: Any, response: Any):
+        local_client = magic_link.Client(
+            db=self.db,
+            tenant=self.tenant,
+            test_mode=self.test_mode,
+            issuer=self.base_path,
+        )
+        query = urllib.parse.parse_qs(
+            request.url.query.decode("ascii") if request.url.query else ""
+        )
+        token = _get_search_param(query, "token")
+
+        (identity_id, challenge, callback_url) = (
+            self._get_data_from_magic_link_token(token)
```

Comment:
We should probably catch all the errors in this endpoint (like the expired token error), and always redirect, since this is a link that the user follows from an email and not an api call, so we don't want to return json.

Reply:
oh, that's tricky since we don't know where to redirect errors to. I suppose we'll need to pass that through as part of the URL search params. Or maybe the redirect error should be configured in the provider config? 🤔 But we'll want to use the built-in UI _if_ you're coming from the built-in UI, so maybe that means search params are the way to go there.

Furthermore, if the redirect or token is missing from the search params, I guess we'll need to render an HTML error page?

Reply:
Yeah, I guess we do all of this in the reset password handler, so I'll just follow that same pattern here.